zkevm-circuits/src/pi_circuit.rs

```diff
@@ -256,113 +393,276 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
+                q_digest_value_start_cur.clone() * (1.expr() - q_digest_value_start_cur),
+                is_value_rlc_cur.clone() * (1.expr() - is_value_rlc_cur),
+            ])]
         });
 
-        let offset = BLOCK_LEN + 1 + EXTRA_LEN;
-        let tx_table_len = max_txs * TX_LEN + 1;
+        // 4: rpi_value_rlc[i+1] = rpi_value_rlc[i] * (is_value_rlc[i] ? evm_rand : byte_pow_base )
+        // + rpi_bytes[i+1]
+        meta.create_gate(
+            "rpi_value_rlc[i+1] = rpi_value_rlc[i] * (is_value_rlc[i] ? evm_rand : byte_pow_base) + rpi_bytes[i+1]",
+            |meta| {
+                let mut cb = BaseConstraintBuilder::default();
+                let q_rpi_byte_not_end_cur = meta.query_selector(q_rpi_byte_not_end);
+                let q_rpi_value_start_next = meta.query_fixed(q_rpi_value_start, Rotation::next());
+                let is_value_rlc_cur = meta.query_fixed(is_value_rlc, Rotation::next());
+                let rpi_value_rlc_next = meta.query_advice(rpi_value_rlc, Rotation::next());
+                let rpi_value_rlc_cur = meta.query_advice(rpi_value_rlc, Rotation::cur());
+                let rpi_bytes_next = meta.query_advice(rpi_bytes, Rotation::next());
+                let r = select::expr(is_value_rlc_cur, challenges.evm_word(), BYTE_POW_BASE.expr());
+
+                cb.require_equal(
+                    "rpi_value_rlc[i+1] = rpi_value_rlc[i]* r + rpi_bytes[i+1]",
+                    rpi_value_rlc_next,
+                    rpi_value_rlc_cur * r + rpi_bytes_next,
+                );
 
-        //  0.3 Tx table -> {tx_id, index, value} column match with raw_public_inputs
-        // at expected offset
+                cb.gate(not::expr(q_rpi_value_start_next) * q_rpi_byte_not_end_cur)
+            },
+        );
+
+        // 5. rpi_value_rlc[i] = rpi_bytes[i] if q_rpi_value_start = 1
         meta.create_gate(
-            "tx_table.tx_id[i] == raw_public_inputs[offset + i]",
+            "rpi_value_rlc[i] = rpi_bytes[i] if q_rpi_value_start = 1",
             |meta| {
-                // row.q_tx_table * row.tx_table.tx_id
-                // == row.q_tx_table * row_offset_tx_table_tx_id.raw_public_inputs
-                let q_tx_table = meta.query_selector(q_tx_table);
-                let tx_id = meta.query_advice(tx_table.tx_id, Rotation::cur());
-                let rpi_tx_id = meta.query_advice(raw_public_inputs, Rotation(offset as i32));
+                let q_rpi_value_start_cur = meta.query_fixed(q_rpi_value_start, Rotation::cur());
+                let q_rpi_byte_not_end_cur = meta.query_selector(q_rpi_byte_not_end);
+                let rpi_bytes_cur = meta.query_advice(rpi_bytes, Rotation::cur());
+                let rpi_value_rlc_cur = meta.query_advice(rpi_value_rlc, Rotation::cur());
 
-                vec![q_tx_table * (tx_id - rpi_tx_id)]
+                vec![
+                    q_rpi_byte_not_end_cur
```

Comment:
Is `q_rpi_byte_not_end_cur` really needed here?

Reply:
this field is removed by latest refactor